code/core/system/proxy-velodyne32/include/velodyne32Decoder.h

```diff
@@ -0,0 +1,139 @@
+    uint8_t m_CPCIntensityOption; //Only used when CPC is enabled. 0: without intensity; 1: with intensity; 2: send a CPC container twice, one with intensity, and the other without intensity
+    uint8_t m_numberOfBitsForIntensity; //Range 0-7. Only used when CPC is enabled
+    uint8_t m_intensityPlacement;  //0: higher bits; 1: lower bits
+    uint16_t m_mask;  //for combining distance and intensity in 16 bits
+    uint8_t m_distanceEncoding; //0: cm; 1: 2mm
+    uint32_t m_pointIndexSPC; //current number of points of the current frame for shared point cloud 
+    uint32_t m_pointIndexCPC; //current number of points of the current frame for compact point cloud
+    uint32_t m_startID;
+    float m_previousAzimuth;
+    float m_currentAzimuth;
+    float m_distance;
+    std::shared_ptr< SharedMemory > m_velodyneSharedMemory; //shared memory for shared point cloud
+    float *m_segment;  //temporary memory for transferring data of each frame to the shared memory
+    odcore::io::conference::ContainerConference &m_conference;
+    odcore::data::SharedPointCloud m_spc; //shared point cloud
+    float m_verticalAngle[32];           //Vertical angle of each sensor beam
+    string m_calibration;  //name of the calibration file for HDL-32E
+    const float toRadian = static_cast<float>(M_PI) / 180.0f;  //degree to radian
+    bool m_withSPC;  //if SPC is expected
+    bool m_withCPC;  //if CPC is expected
+        
+    //For compact point cloud:
+    float m_startAzimuth;
+    const uint8_t m_ENTRIES_PER_AZIMUTH = 11;//For HDL-32E, there are 32 points per azimuth
+    std::stringstream m_distanceStringStreamNoIntensityPart1; //Layer 0, 1, 4, 7..., i.e., in addition to Layer 0, every 3rd layer from Layer 1 and resulting in 12 layers, without intensity
+    std::stringstream m_distanceStringStreamNoIntensityPart2; //Layer 2, 3, 6, 9..., i.e., in addition to Layer 2, every 3rd layer from Layer 3 and resulting in 11 layers, without intensity
+    std::stringstream m_distanceStringStreamNoIntensityPart3; //Layer 5, 8, 11..., i.e., every 3rd layer from Layer 5 and resulting in 9 layers, without intensity
+    std::stringstream m_distanceStringStreamWithIntensityPart1; //Similar to m_distanceStringStreamNoIntensityPart1, with intensity
+    std::stringstream m_distanceStringStreamWithIntensityPart2; //Similar to m_distanceStringStreamNoIntensityPart2, with intensity
+    std::stringstream m_distanceStringStreamWithIntensityPart3; //Similar to m_distanceStringStreamNoIntensityPart3, with intensity
+    bool m_isStartAzimuth;  //Indicate if an azimuth is the starting azimuth of a new frame
+    uint8_t m_sensorOrderIndex[32];//Specify the sensor ID order for each 32 points with increasing vertical angle for CPC and SPC
+    uint16_t m_32SensorsNoIntensity[32];//Store the distance values of the current 32 sensors for CPC without intensity
+    uint16_t m_32SensorsWithIntensity[32];//Store the distance values of the current 32 sensors for CPC with intensity
```

Comment:
Please use std::array over the old C-style arrays as they are type-safe and provide generic STL-container-like access: http://en.cppreference.com/w/cpp/container/array